nat-lab/bin/netcat.py

```diff
@@ -0,0 +1,209 @@
+            sys.exit(1)
+
+    def _connect(self):
+        """Connect to a remote listener"""
+        hostname = self.args.hostname
+        port = self.args.port
+
+        try:
+            # Bind to a local port or source address if specified
+            if self.args.p or self.args.s:
+                local_address: str = self.args.s if self.args.s else ""
+                local_port: int = self.args.p if self.args.p else 0
+
+                self.sock.bind((local_address, local_port))
+
+            self.sock.connect((hostname, port))
+            if self.udp:
+                self._udptest()
+            self._vprint(
+                f"Connection to {hostname} {port} port [{self.sock_type}/*] succeeded!"
+            )
+        except OSError as e:
+            if e.errno == errno.EBADF:
+                # connection closed
+                sys.exit(0)
+            print(
+                f"nc: connect to {hostname} port {port} ({self.sock_type}) failed: {e.strerror}",
+                file=sys.stderr,
+            )
+            sys.exit(1)
+
+    def _listen(self):
+        """Listen for incoming connections"""
+        # Set SO_REUSEPORT option
+        try:
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+        except AttributeError:
+            # SO_REUSEPORT may not be available on all systems, fall back to SO_REUSEADDR
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+        port = self.args.port
+        bind_addr = "::" if self.ipv6 else "0.0.0.0"
+        self.sock.bind((bind_addr, port))
+
+        if self.udp:
+            self._vprint(f"Bound on {bind_addr} {port}")
+        else:
+            self.sock.listen(1)
+            self._vprint(f"Listening on {bind_addr} {port}")
+            conn, addr = self.sock.accept()
+            self._vprint(f"Connection received on {addr[0]} {addr[1]}")
+            self.sock = conn
+
+    def _udptest(self):
+        """Test UDP connection"""
+        # Try sending data to the socket
+        for _ in range(UDP_SCAN_COUNT):
+            self.sock.send(b"X")
+            time.sleep(1)
+
+    def _register_socket(self):
+        """Register the socket to the selector for read events"""
+        self.selector.register(self.sock, selectors.EVENT_READ, self._read_from_socket)
+
+    def _register_stdin(self):
+        """Register stdin to the selector for read events, unless disabled"""
+        if not self.args.d:
+            self.selector.register(
+                sys.stdin, selectors.EVENT_READ, self._read_from_stdin
+            )
+
+    def _read_from_socket(self):
+        """Handle incoming data from the socket"""
+        data, addr = self.sock.recvfrom(RECV_SIZE)
+        if self.udp and self.listen and not self.client_addr:
+            self._vprint(f"Connection received on {addr[0]} {addr[1]}")
+            self.client_addr = addr
+            self.sock.connect(addr)
```

Comment:
I guess for our purposes -> I am not sure if we really need connected sockets. Maybe it is worth simplifying our nc variant, and just always use unconnected sockets for UDP?

Reply:
Hmm, I tried here to implement [this bit](https://sources.debian.org/src/netcat-openbsd/1.195-2/netcat.c/#L774), without the `connect()` the process will not finish when trying to send data if the listener is gone.

Reply:
You're right, this is one is an extra here, I probably forgot it when refactoring things

Reply:
good point about exiting. Missed it.